core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryStream.scala

```diff
@@ -0,0 +1,146 @@
+package com.softwaremill.kmq.redelivery.streams
+
+import akka.Done
+import akka.actor.ActorSystem
+import akka.kafka.ConsumerMessage.CommittableMessage
+import akka.kafka.scaladsl.Consumer
+import akka.kafka.scaladsl.Consumer.DrainingControl
+import akka.kafka.{ConsumerSettings, Subscriptions}
+import akka.stream.scaladsl.{Flow, Keep, Sink, Source}
+import com.softwaremill.kmq._
+import com.softwaremill.kmq.redelivery.streams.RedeliveryStream._
+import com.softwaremill.kmq.redelivery.{DefaultRedeliverer, Partition, RetryingRedeliverer, Timestamp}
+import com.typesafe.scalalogging.{Logger, StrictLogging}
+import org.apache.kafka.common.serialization.ByteArraySerializer
+
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+
+class RedeliveryStream(markerConsumerSettings: ConsumerSettings[MarkerKey, MarkerValue],
+                       markersTopic: String, maxPartitions: Int,
+                       kafkaClients: KafkaClients, kmqConfig: KmqConfig)
+                      (implicit system: ActorSystem) extends StrictLogging {
+
+  private val producer = kafkaClients.createProducer(classOf[ByteArraySerializer], classOf[ByteArraySerializer])
+
+  def redeliverySink(partition: Partition): Sink[CommittableMessage[MarkerKey, MarkerValue], Future[Done]] = {
+    Flow[CommittableMessage[MarkerKey, MarkerValue]]
+      .map(MarkerRedeliveryCommand)
+      .merge(Source.tick(initialDelay = 1.second, interval = 1.second, tick = TickRedeliveryCommand))
+      .statefulMapConcat { () => // keep track of open markers; select markers to redeliver
+        val markersByTimestamp = new PriorityQueueMap[MarkerKey, MsgWithTimestamp](valueOrdering = bySmallestTimestampAscending)
+        val latestMarkerSeenTimestamp = new Ref[Timestamp]
+        val latestMarkerSeenAt = new Ref[Timestamp]
+        cmd => {
+          logger.traceCommand(cmd)
+          cmd match {
+            case TickRedeliveryCommand => // nothing to do
+            case MarkerRedeliveryCommand(msg) =>
+              // update markersByTimestamp
+              msg.record.value match {
+                case _: StartMarker => markersByTimestamp.put(msg.record.key, MsgWithTimestamp(msg, msg.record.timestamp + msg.record.value.asInstanceOf[StartMarker].getRedeliverAfter)) //TODO: simplify
```

Comment:
instead of the cast you can do `case start: StartMarker` and use `start` in the body